runtime/common/src/main/kotlin/kotlinx/serialization/json/serializers/JsonElementSerializer.kt

```diff
@@ -68,13 +67,68 @@ private object JsonObjectSerializer : KSerializer<JsonObject> {
     }
 }
 
-private object JsonPrimitiveSerializer : KSerializer<JsonPrimitive> {
+@Serializer(forClass = JsonPrimitive::class)
+public object JsonPrimitiveSerializer : KSerializer<JsonPrimitive> {
 
     override fun serialize(encoder: Encoder, obj: JsonPrimitive) {
-        if (obj is JsonNull) {
-            return encoder.encodeString(obj.content)
+        return if (obj is JsonNull) {
+            JsonNullSerializer.serialize(encoder, JsonNull)
+        } else {
+            JsonLiteralSerializer.serialize(encoder, obj as JsonLiteral)
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): JsonPrimitive {
+        val nullable = decoder.decodeNullable(StringSerializer)
+        return if (nullable == null) JsonNull else JsonPrimitive(nullable)
```

Comment:
Hm, I came up with a better idea: we have `decodeNotNullMark` which says if next element is null. You can check it and call `JsonNullSerializer.deserialize(decoder)` or `JsonLiteral.deserialize(decoder)`.